### PR TITLE
Add permission to secure access to certain config files specified by settings

### PR DESCRIPTION
**Pull request by thecoop**
Follow on to #107827, adding a setting to specify config files in settings rather than direct

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/SecuredFileSettingAccessPermission.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+package org.elasticsearch;
+
+import java.security.BasicPermission;
+
+/**
+ * A permission granted to ensure secured access to a file specified by a setting in the config directory.
+ * <p>
+ * By granting this permission with a setting key (wildcards are supported),
+ * the specified setting(s) are read to specify files in the config directory that should be secured.
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/SecuredFileSettingAccessPermission.java:17]: This wording is a little confusing, perhaps something like this:
```
the file pointed to by the setting is secured from general access by other Elasticsearch plugins.
The file path is relative to the config directory.
```

File: server/src/main/java/org/elasticsearch/bootstrap/Security.java [hunk 2]
```diff
@@ -122,6 +127,8 @@ static void setSecurityManager(@SuppressWarnings("removal") SecurityManager sm)
      * @param filterBadDefaults true if we should filter out bad java defaults in the system policy.
      */
     static void configure(Environment environment, boolean filterBadDefaults, Path pidFile) throws IOException {
+        Log = LogManager.getLogger(Security.class);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/bootstrap/Security.java:130]: nit, use `logger` for consistency with other code, even if just a local?

File: server/src/main/java/org/elasticsearch/SecuredFileSettingAccessPermission.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ * All code that does not have the same permission on the same file will be denied all read/write access to that file.
+ * Note that you also need to wrap any access to the secured files in an {@code AccessController.doPrivileged()} block
+ * as Elasticsearch itself is denied access to files secured by plugins.
+ */
+public class SecuredFileSettingAccessPermission extends BasicPermission {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/SecuredFileSettingAccessPermission.java:22]: Perhaps the name should indicate it is a config file that is being pointed to?

File: server/src/main/java/org/elasticsearch/bootstrap/Security.java [hunk 4]
```diff
@@ -203,21 +210,39 @@ private static Map<String, Set<URL>> readSecuredFiles(
...
+        List<Map.Entry<Pattern, Set<URL>>> settingPatterns = securedSettingKeys.entrySet()
+            .stream()
+            .map(e -> Map.entry(Pattern.compile(e.getKey()), e.getValue()))
+            .toList();
+
+        for (String setting : environment.settings().keySet()) {
+            for (Map.Entry<Pattern, Set<URL>> ps : settingPatterns) {
+                if (ps.getKey().matcher(setting).matches()) {
+                    // add the setting value to the secured files for these codebase URLs
+                    Path file = environment.configFile().resolve(environment.settings().get(setting));
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/bootstrap/Security.java:238]: We should make sure the file is actually rooted in the config directory, ie a relative path wasn't specified that traverses outside it

File: server/src/main/java/org/elasticsearch/bootstrap/Security.java [hunk 5]
```diff
@@ -230,21 +255,41 @@ private static Map<String, Set<URL>> readSecuredFiles(
...
+    private static void readSecuredPermissions(
+        Environment environment,
+        URL url,
+        Permission p,
+        Map<String, Set<URL>> securedFiles,
+        Map<String, Set<URL>> securedSettingKeys
+    ) {
+        String securedFileName = extractSecuredName(p, SecuredFileAccessPermission.class);
+        if (securedFileName != null) {
+            Path securedFile = environment.configFile().resolve(securedFileName);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/bootstrap/Security.java:267]: Same here, let's make sure this is actually rooted in the config directory

**Review by rjernst — COMMENTED** [review]
Looks good, one general comment about ensuring paths are actually config files.

File: server/src/main/java/org/elasticsearch/SecuredFileSettingAccessPermission.java (hunk 3 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/SecuredFileSettingAccessPermission.java:22]: I'm just concerned about the name being too long - `SecuredConfigFileSettingAccessPermission`

File: server/src/main/java/org/elasticsearch/bootstrap/Security.java (hunk 4 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/bootstrap/Security.java:238]: I've added a subpath check

File: server/src/main/java/org/elasticsearch/bootstrap/Security.java (hunk 5 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/bootstrap/Security.java:272]: I've added a subpath check

File: server/src/main/java/org/elasticsearch/SecuredFileAccessPermission.java [hunk 6]
```diff
@@ -13,9 +13,10 @@
...
- * By granting this permission, all code that does not have the same permission on the same file
- * will be denied all read/write access to that file.
- * Note that you also need to wrap any access to the secured files in an {@code AccessController.doPrivileged()} block
+ * By granting this permission with a file relative to the config directory,
+ * the file is secured from general access by Elasticsearch and other Elasticsearch plugins.
+ * All code that does not have a secured permission on the same file will be denied all read/write access to that file.
+ * Note that you also need to wrap any access to secured files in an {@code AccessController.doPrivileged()} block
  * as Elasticsearch itself is denied access to files secured by plugins.
  */
 public class SecuredFileAccessPermission extends BasicPermission {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/SecuredFileAccessPermission.java:22]: This should also have "config" in the name?

File: server/src/main/java/org/elasticsearch/bootstrap/Security.java [hunk 7]
```diff
@@ -230,21 +260,44 @@ private static Map<String, Set<URL>> readSecuredFiles(
         return Collections.unmodifiableMap(securedFiles);
     }
 
-    private static void addSpeciallySecuredFile(Map<String, Set<URL>> securedFiles, String path) {
-        Set<URL> attemptedToGrant = securedFiles.put(path, Set.of());
-        if (attemptedToGrant != null) {
-            throw new IllegalStateException(attemptedToGrant + " tried to grant access to special config file " + path);
+    private static void readSecuredPermissions(
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/bootstrap/Security.java:263]: nit: readSecuredConfigFilePermissions?

**Review by rjernst — APPROVED** [review]
Looks good, assuming you agree with clarity of names.

**Comment by thecoop** [comment]
@elasticsearch rerun elasticsearch-ci/part-3

**Comment by thecoop** [comment]
@elasticmachine rerun elasticsearch-ci/part-4-fips

**Comment by thecoop** [comment]
Test failures also happen on main, merging

**Pull request merged**